### ISSUE TITLE
The processing procedure of thermal coding in the loss function of object detection

### ISSUE THREAD

**Issue by cainiao123s** [ultralytics/ultralytics#20669]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Firstly, the thermal coding has been completed through the following code：
        # 10x faster than F.one_hot()
        target_scores = torch.zeros(
            (target_labels.shape[0], target_labels.shape[1], self.num_classes),
            dtype=torch.int64,
            device=target_labels.device,
        )  # (b, h*w, 80)
        target_scores.scatter_(2, target_labels.unsqueeze(-1), 1)
However, the background anchor points are taken to 0 through the mask. I don't understand why this operation is performed. In one-hot encoding, if the anchor points do not contain the target, then all categories have already been set to 0. So is the following code a check operation? Or does it have other meanings?
        fg_scores_mask = fg_mask[:, :, None].repeat(1, 1, self.num_classes)  # (b, h*w, 80)
        target_scores = torch.where(fg_scores_mask > 0, target_scores, 0)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @cainiao123s, thank you for your interest in Ultralytics 🚀! This is an automated response to help guide your issue—an Ultralytics engineer will review your question and assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for detailed explanations and to explore [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions about loss functions and label processing are already covered there.

If your question relates to a 🐛 Bug Report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) to help us understand and debug your scenario.

For custom training or implementation ❓ Questions, please share as much detail as possible—such as code snippets, dataset image examples, or training logs. Also, check our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure optimal outcomes.

Join the Ultralytics community wherever you prefer! For real-time chat, join us on [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth discussions, visit [Discourse](https://community.ultralytics.com/). Or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics).

## Upgrade

Make sure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models can be run in any of these up-to-date, verified environments with all dependencies (including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/)) preinstalled:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for contributing and sharing your question!

**Comment by glenn-jocher** [comment]
Thank you for your question regarding the thermal coding in the loss function. 

You're looking at the `get_targets` function from `ultralytics/utils/tal.py`. The operation you're asking about is indeed necessary and serves an important purpose.

When the code creates `target_scores` using the scatter operation, it assigns one-hot encoded values for *all* anchor points based on their assigned target labels. However, not all anchor points are actually foreground (positive) anchors.

The `fg_mask` specifically indicates which anchor points are truly foreground (positive), and the final operation:
```python
fg_scores_mask = fg_mask[:, :, None].repeat(1, 1, self.num_classes)  # (b, h*w, 80)
target_scores = torch.where(fg_scores_mask > 0, target_scores, 0)
```

ensures that only those anchor points marked as foreground in the `fg_mask` keep their one-hot encoded scores. All other anchor points (the background ones) are explicitly set to zero across all classes.

This isn't just a check operation - it's essential for ensuring that background anchor points don't contribute to the classification loss, even though they have been assigned labels in the earlier steps of the task-aligned assignment process.

**Comment by cainiao123s** [comment]
So the values of target-scores won't change before and after processing. Is my conclusion correct?



---Original---
From: "Glenn ***@***.***&gt;
Date: Sat, May 17, 2025 03:00 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [ultralytics/ultralytics] The processing procedure of thermalcoding in the loss function of object detection (Issue #20669)


 glenn-jocher left a comment (ultralytics/ultralytics#20669)
 
Thank you for your question regarding the thermal coding in the loss function.
 
You're looking at the get_targets function from ultralytics/utils/tal.py. The operation you're asking about is indeed necessary and serves an important purpose.
 
When the code creates target_scores using the scatter operation, it assigns one-hot encoded values for all anchor points based on their assigned target labels. However, not all anchor points are actually foreground (positive) anchors.
 
The fg_mask specifically indicates which anchor points are truly foreground (positive), and the final operation:
 fg_scores_mask = fg_mask[:, :, None].repeat(1, 1, self.num_classes)  # (b, h*w, 80) target_scores = torch.where(fg_scores_mask &gt; 0, target_scores, 0)
 
ensures that only those anchor points marked as foreground in the fg_mask keep their one-hot encoded scores. All other anchor points (the background ones) are explicitly set to zero across all classes.
 
This isn't just a check operation - it's essential for ensuring that background anchor points don't contribute to the classification loss, even though they have been assigned labels in the earlier steps of the task-aligned assignment process.

**Comment by cainiao123s** [comment]
> fg_scores_mask = fg_mask[

I have verified through the code that the target_scores have changed.

**Comment by glenn-jocher** [comment]
You're absolutely right. The values of `target_scores` do change during that operation. 

Let me clarify what happens step by step:

1. First, the scatter operation assigns one-hot encoded values (1s) to corresponding class positions for *all* anchor points based on `target_labels`
2. Then the `torch.where()` operation with the `fg_mask` filtering sets scores for all background anchor points (those where `fg_mask` is 0) back to zeros

So the operation doesn't just validate - it actively filters out scores for background anchor points, ensuring only foreground anchors contribute to the classification loss. Background anchors that initially received one-hot encodings in the scatter step get reset to zeros in this step.

This is a crucial part of how YOLO's task-aligned label assignment works to focus the model's learning on relevant foreground objects.

**Comment by cainiao123s** [comment]
> Thank you for your question regarding the thermal coding in the loss function.
> 
> You're looking at the `get_targets` function from `ultralytics/utils/tal.py`. The operation you're asking about is indeed necessary and serves an important purpose.
> 
> When the code creates `target_scores` using the scatter operation, it assigns one-hot encoded values for _all_ anchor points based on their assigned target labels. However, not all anchor points are actually foreground (positive) anchors.
> 
> The `fg_mask` specifically indicates which anchor points are truly foreground (positive), and the final operation:
> 
> fg_scores_mask = fg_mask[:, :, None].repeat(1, 1, self.num_classes)  # (b, h*w, 80)
> target_scores = torch.where(fg_scores_mask > 0, target_scores, 0)
> ensures that only those anchor points marked as foreground in the `fg_mask` keep their one-hot encoded scores. All other anchor points (the background ones) are explicitly set to zero across all classes.
> 
> This isn't just a check operation - it's essential for ensuring that background anchor points don't contribute to the classification loss, even though they have been assigned labels in the earlier steps of the task-aligned assignment process.

In fact, what I want to ask is that when the annotation was made, it was already determined whether this anchor point was a positive sample. However, this function re-filtered the positive samples once again. I don't quite understand the reason for doing this

**Comment by glenn-jocher** [comment]
You're asking an excellent question about the process. To clarify: the positive samples aren't fully determined just at annotation time. 

The full determination of which anchor points are "positive" happens through the Task-Aligned Assignment process during training. Here's the key distinction:

1. Annotations provide the ground truth boxes and labels 
2. The model dynamically determines which anchor points should be responsible for detecting each object using metrics like overlaps and predicted scores
3. This assignment happens in the `TaskAlignedAssigner` class, which calculates the `fg_mask` based on multiple factors

So when you reach the `get_targets` function, the `fg_mask` represents the result of this complex assignment process. The filtering with `fg_mask` is necessary because the scatter operation initially assigns labels to all anchors based on their assigned ground truth, but we only want anchors that the assigner has determined are truly "positive" for that iteration to contribute to the loss.

This dynamic assignment is at the core of YOLO's training approach and improves performance compared to using fixed positive/negative assignments.

**Comment by cainiao123s** [comment]
Hello, I would like to ask, before target_labels is filtered, since it is directly generated based on the label box, does it necessarily present a regular rectangle? For example, as shown in the following figure, it contains the splicing of two rectangles. After filtering, target_labels may become another graph.
![Image](https://github.com/user-attachments/assets/53267d0a-9f24-4987-ae5a-0af75525a9b6)

![Image](https://github.com/user-attachments/assets/79479606-ef87-4e71-b7fa-348dd99f74f7)

**Comment by glenn-jocher** [comment]
You're asking an insightful question about the spatial distribution of positive anchors. 

To clarify - even before filtering with `fg_mask`, the positive anchor points don't necessarily form a perfect rectangle matching the ground truth box. The Task-Aligned Assigner selects positive anchors based on multiple criteria:

1. First with `mask_in_gts` to identify anchors inside ground truth boxes
2. Then calculating alignment metrics between anchors and ground truths 
3. Finally selecting top-k anchors based on these metrics

So when two objects are close (as in your second image), the pre-filtered positive anchors already have a spatial distribution that follows the assignment logic, not simply the rectangular ground truth boxes.

The `fg_mask` filtering is the final step that confirms which anchors are truly "positive" for loss calculation. The resulting spatial pattern of positive anchors can indeed be irregular and is optimized for better object detection performance rather than maintaining geometric shapes.

This dynamic assignment approach is why YOLO performs well even with overlapping objects or complex scenes.

**Comment by cainiao123s** [comment]
> 您问了一个关于正锚点的空间分布的深刻问题。
> 
> 需要澄清的是，即使在使用 进行过滤之前`fg_mask`，正锚点也不一定能形成与真实框完美匹配的矩形。任务对齐分配器会根据多个标准选择正锚点：
> 
> 1. 首先`mask_in_gts`识别真实框内的锚点
> 2. 然后计算锚点和真实值之间的对齐指标
> 3. 最后根据这些指标选择 top-k 锚点
> 
> 因此，当两个物体靠近时（如第二张图像所示），预过滤的正锚点已经具有遵循分配逻辑的空间分布，而不仅仅是矩形的地面真实框。
> 
> 过滤`fg_mask`是最后一步，用于确认哪些锚点对于损失计算而言是真正的“正”锚点。由此产生的正锚点的空间模式确实可以是不规则的，并且是为了获得更好的物体检测性能而进行优化，而不是维持几何形状。
> 
> 这种动态分配方法是 YOLO 即使在重叠物体或复杂场景下也能表现良好的原因。

Let me try to explain my understanding.
The function of mask_in_gts = self.select_candidates_in_gts(anc_points, gt_bboxes) is to filter out the anchor points (anc_points) within the real boxes (gt_bboxes).
Therefore, the shape of mask_in_gts should be the distribution form in Figure 1
target_scores are initially generated by the following code:
target_scores = torch.zeros(
(target_labels.shape[0], target_labels.shape[1], self.num_classes),
dtype=torch.int64,
device=target_labels.device,
So I think whether each anchor point contains the target (if it does not contain the target, the category scores are all 0), target_scores should correspond one-to-one with mask_in_gts.

**Comment by glenn-jocher** [comment]
You're absolutely correct in your understanding! 

The `mask_in_gts` from `select_candidates_in_gts` does indeed filter anchor points that fall within the ground truth bounding boxes, creating the rectangular distribution pattern you showed in Figure 1.

However, there's an important distinction: `target_scores` is generated based on `target_labels`, which comes from the final assignment after the full Task-Aligned Assignment process. This process includes additional filtering beyond just `mask_in_gts`:

1. `mask_in_gts` - anchors inside GT boxes (rectangular pattern)
2. `mask_topk` - top-k anchors based on alignment metrics  
3. `mask_pos = mask_topk * mask_in_gts * mask_gt` - final positive mask

So `target_scores` doesn't correspond one-to-one with `mask_in_gts` alone, but rather with the final `fg_mask` that results from this multi-step assignment process. The `fg_mask` filtering in `get_targets` ensures that only the anchors that survived the complete assignment process retain their classification targets, while others are set to zero.

This is why the spatial pattern can be more complex than the simple rectangular `mask_in_gts` pattern you described.
